rfcs/react-components/convergence/event-handlers-event-type.md

```diff
@@ -0,0 +1,198 @@
+# RFC: Extending Event Handlers Event Typing
+
+This RFC proposes solutions to enhance event typing within callbacks.
+
+---
+
+@yuanboxue-amber
+
+## Summary
+
+We encounter challenges when attempting to extend event types within callbacks without introducing breaking changes. This RFC offers potential solutions to manage such event type changes for v9 components.
+Notably, this RFC explores alternatives for v10 in discarded solutions, but it is not aiming for proposing a v10 callback type design.
+
+## Problem Statement
+
+Our current design for event callbacks in v9 has the type signature `onSomeEvent: (e: SomeEvents, data: SomeData) => void;`, where `SomeEvents` represents specific event types like MouseEvent or KeyboardEvent, or a union of specific event types. While our intention is to maintain strong type checking, this design complicates the process of expanding or altering event types without causing breaks.
+
+We experienced this challenge when when adding a scroll event listener to the Popover `onOpenChange` callback. The callback has type `onOpenChange?: (e: OpenPopoverEvents, data: OnOpenChangeData) => void;`, where `OpenPopoverEvents` is a union of specific events. But the scroll event has just type `Event`. And we are facing issues when updating the type:
+
+```ts
+export type OpenPopoverEvents = KeyboardEvent | MouseEvent;
+type Props = {
+  onOpenChange: (e: OpenPopoverEvents) => void;
+};
+const onOpenChange = (e: MouseEvent | KeyboardEvent) => {};
+const props: Props = { onOpenChange }; // 💣 This breaks when we add Event to OpenPopoverEvents. The error: '(e: MouseEvent | KeyboardEvent) => void' is not assignable to type '(e: OpenPopoverEvents) => void'.
+```
+
+We cannot just add the `Event` type to the existing callbacks as this is a breaking change for any code that explicitly types its event argument. This is a challenge especially when unforeseen events need to be added in the future.
+
+## Detailed Proposal
+
+Use generic [`Event`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/v17/global.d.ts#L10) type and strongly-typed events in "data bag".
+
+### Implementation
+
+Define two helper types `EventData` and `EventHandler` to ensure consistency across all components:
+
+```ts
+export type EventData<Type extends string, TEvent> =
+  | { type: undefined; event: React.SyntheticEvent | Event | undefined }
+  | { type: Type; event: TEvent };
+
+export type EventHandler<TData extends EventData<string, unknown>> = (
+  ev: React.SyntheticEvent | Event | undefined,
+  data: TData,
+) => void;
+```
+
+- For existing event handlers where we can't change the type:
+
+  - Deprecate the existing `onSomeEvent` handler and introduce `onSomeEvent2` with the new proposed type signatures.
+  - Continue to call the deprecated `onSomeEvent` in all cases. If the event object is of the wrong type for the existing function, do a cast.
```

Comment:
 We could potentially add a `LegacyEventHandler` type existing handlers. It's not strictly necessary, but it:
1. Ensures consistency that all Data types include `EventData` (by the restriction that `TData extends EventData`).
2. Makes it clear(er) that new components shouldn't copy this "legacy" pattern.
3. If we choose to update the types in v10, this will make it slightly easier to do that (replace `LegacyEventHandler` with `EventHandler`).

```ts
export type LegacyEventHandler<TEvent, TData extends EventData<string, TEvent>> = (
  ev: TEvent,
  data: TData,
) => void;
```